qos/cometbft/context.go

```diff
@@ -0,0 +1,164 @@
+package cometbft
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/pokt-network/poktroll/pkg/polylog"
+
+	"github.com/buildwithgrove/path/gateway"
+	qosobservations "github.com/buildwithgrove/path/observation/qos"
+	"github.com/buildwithgrove/path/protocol"
+)
+
+const defaultServiceRequestTimeoutMillisec = 10_000
+
+// requestContext provides the support required by the gateway
+// package for handling service requests.
+var _ gateway.RequestQoSContext = &requestContext{}
+
+type response interface {
+	GetObservation() qosobservations.CometBFTEndpointObservation
+	GetResponsePayload() []byte
+	GetResponseStatusCode() int
+}
+
+type endpointResponse struct {
+	protocol.EndpointAddr
+	response
+	unmarshalErr error
+}
+
+// requestContext provides the functionality required
+// to support QoS for a CometBFT blockchain service.
+type requestContext struct {
+	logger polylog.Logger
+
+	httpReq       *http.Request
+	endpointStore *EndpointStore
+
+	// isValid indicates whether the underlying user request
+	// for this request context was found to be valid.
+	// This field is set by the corresponding QoS instance
+	// when creating this request context during the parsing
+	// of the user request.
+	isValid bool
+
+	// preSelectedEndpointAddr allows overriding the default
+	// endpoint selector with a specific endpoint's addresss.
+	// This is used when building a request context as a check
+	// for a specific endpoint.
+	preSelectedEndpointAddr protocol.EndpointAddr
+
+	// endpointResponses is the set of responses received from one or
+	// more endpoints as part of handling this service request.
+	// NOTE: these are all related to a single JSONRPC request,
+	// enhancing to support batch JSONRPC requests will involve the
+	// modification of this field's type.
+	endpointResponses []endpointResponse
+}
+
+func (rc requestContext) GetServicePayload() protocol.Payload {
+	return protocol.Payload{
+		//CometBFT uses only GET HTTP requests with empty body
+		Data: "",
+
+		// Method is alway GET for CometBFT-based blockchains.
```

Comment:
Looks like this is why we are not using an HTTP method elsewhere.
Not a blocker, but I do recommend mentioning this (and referencing this PR in a TODO)

Reply:
> Looks like this is why we are not using an HTTP method elsewhere. Not a blocker, but I do recommend mentioning this (and referencing this PR in a TODO)

This actually alerted me to an important catch: the CometBFT spec actually is intended to support both REST-like requests (eg `GET /status`) and JSON-RPC requests (eg `POST / -d "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"status\"}"`.

I've update the `cometbft` package to reflect this so that both types of requests will work.